### PR TITLE
[Shallow] New element

### PR DESCRIPTION
**Pull request by miguelmaso**
**Description**
This PR adds an element implementing SUPG stabilization for the conservative SW equations. This is a partial step for:
- Implement a GJV shock capturing to improve monotonicity.
- Derive the current standard element from the new implementation and avoid a lot of code duplication.

**Changelog**
- Update the `WaveElement` as base element to avoid code duplication.
- Add `ConservativeElement` derived from `WaveElement`.
- Apply the corresponding changes to the conditions.
- Add tests utilities to avoid code duplication.
- Add tests.
- Minor updates on `CrankNicolsonWaveElement` and `SWE`.

### CODE REVIEW

**Review by ipouplana — DISMISSED** [review]

File: applications/ShallowWaterApplication/custom_conditions/conservative_condition.h [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+    /**
+     * @brief Constructor using Geometry and Properties
+     */
+    ConservativeCondition(IndexType NewId, GeometryType::Pointer pGeometry, typename PropertiesType::Pointer pProperties) : WaveConditionType(NewId, pGeometry, pProperties){}
+
+    /**
+     * @brief Destructor
+     */
+    virtual ~ ConservativeCondition(){}
```
Comment (loumalouomega) [inline applications/ShallowWaterApplication/custom_conditions/conservative_condition.h]: Minor comment, but if this is a condition should override

File: applications/ShallowWaterApplication/custom_conditions/wave_condition.cpp [hunk 2]
```diff
@@ -102,9 +104,9 @@ void WaveCondition<TNumNodes>::GetValuesVector(Vector& rValues, int Step) const
     IndexType counter = 0;
     for (IndexType i = 0; i < TNumNodes; i++)
     {
-        rValues[counter++] = r_geom[i].FastGetSolutionStepValue(VELOCITY_X, Step);
-        rValues[counter++] = r_geom[i].FastGetSolutionStepValue(VELOCITY_Y, Step);
-        rValues[counter++] = r_geom[i].FastGetSolutionStepValue(HEIGHT, Step);
+        rValues[counter++] = r_geom[i].FastGetSolutionStepValue(this->GetUnknownComponent(0), Step);
```
Comment (loumalouomega) [inline applications/ShallowWaterApplication/custom_conditions/wave_condition.cpp:107]: Unfortunately, this is more "general", but less efficient. You can consider a static array, or something else in order to improve efficiency

File: applications/ShallowWaterApplication/custom_elements/conservative_element.cpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+#include "custom_utilities/shallow_water_utilities.h"
+
+namespace Kratos
+{
+
+template<std::size_t TNumNodes>
+const Variable<double>& ConservativeElement<TNumNodes>::GetUnknownComponent(int Index) const
+{
+    switch (Index) {
+        case 0: return MOMENTUM_X;
```
Comment (loumalouomega) [inline applications/ShallowWaterApplication/custom_elements/conservative_element.cpp:32]: Idem as before

File: applications/ShallowWaterApplication/custom_elements/conservative_element.h [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+    /**
+     * @brief Constructor using Geometry and Properties
+     */
+    ConservativeElement(IndexType NewId, GeometryType::Pointer pGeometry, typename PropertiesType::Pointer pProperties) : WaveElementType(NewId, pGeometry, pProperties){}
+
+    /**
+     * @brief Destructor
+     */
+    virtual ~ ConservativeElement(){}
```
Comment (loumalouomega) [inline applications/ShallowWaterApplication/custom_elements/conservative_element.h]: Idem

File: applications/ShallowWaterApplication/custom_utilities/sw_math_utils.h [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    ///@name Operators
+    ///@{
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    template<class TMatrixTypeA, class TMatrixTypeB>
+    static void AddMatrix(
```
Comment (loumalouomega) [inline applications/ShallowWaterApplication/custom_utilities/sw_math_utils.h:58]: This is generic enough to be in the Core math utils

**Review by loumalouomega — COMMENTED** [review]
Some comments

File: applications/ShallowWaterApplication/custom_conditions/conservative_condition.h (hunk 1 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_conditions/conservative_condition.h]: thanks, I'm going to modify it.

File: applications/ShallowWaterApplication/custom_utilities/sw_math_utils.h (hunk 5 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_utilities/sw_math_utils.h:58]: I think the same, just now I'm preparing another PR to move it to the core

File: applications/ShallowWaterApplication/custom_conditions/wave_condition.cpp (hunk 2 above)
Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_conditions/wave_condition.cpp:107]: I've tested the efficiency and the cost is acceptable.

Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_conditions/wave_condition.cpp:107]: How can I define the static array in derived classes? Do you have an example?

Comment (loumalouomega) [reply applications/ShallowWaterApplication/custom_conditions/wave_condition.cpp:107]: You mean that the method is virtual ?, I did not see that, sorry

Comment (loumalouomega) [reply applications/ShallowWaterApplication/custom_conditions/wave_condition.cpp:107]: > I've tested the efficiency and the cost is acceptable.

imagen

**Review by ipouplana — APPROVED** [review]

**Pull request merged**